CyclopsSolarUpgrades/Management/SolarCharger.cs

```diff
@@ -51,69 +53,75 @@ public Atlas.Sprite GetIndicatorSprite()
-                    return NumberFormatter.FormatNumber(Mathf.CeilToInt(solarPercentage), NumberFormat.Sun);
+                    return TextFormatter.FormatSolarPercentage(solarPercentage);
                 case SolarState.BatteryAvailable:
-                    return NumberFormatter.FormatNumber(Mathf.CeilToInt(this.SolarChargerMk2.TotalBatteryCharge), NumberFormat.Amount);
+                    return TextFormatter.FormatBatteryCharge(upgradeHandler.TotalBatteryCharge);
                 default:
                     return string.Empty;
             }
         }
 
         public Color GetIndicatorTextColor()
         {
-            switch (SolarState)
+
+            switch (solarState)
             {
                 case SolarState.SunAvailable:
-                    return NumberFormatter.GetNumberColor(solarPercentage, MaxSolarPercentage, 0f);
+                    return TextFormatter.GetNumberColor(solarPercentage, MaxSolarPercentage, 0f);
                 case SolarState.BatteryAvailable:
-                    return NumberFormatter.GetNumberColor(this.SolarChargerMk2.TotalBatteryCharge, this.SolarChargerMk2.TotalBatteryCapacity, 0f);
+                    return TextFormatter.GetNumberColor(upgradeHandler.TotalBatteryCharge, upgradeHandler.TotalBatteryCapacity, 0f);
                 default:
                     return Color.white;
             }
         }
 
         public bool HasPowerIndicatorInfo()
         {
-            return SolarState != SolarState.None;
+            return solarState != SolarState.None;
         }
 
         public float ProducePower(float requestedPower)
         {
-            if (this.SolarChargers.Count == 0 && this.SolarChargerMk2.Count == 0)
+            if (upgradeHandler.TotalCount == 0)
             {
-                SolarState = SolarState.None;
+                solarState = SolarState.None;
                 return 0f;
             }
 
             float solarStatus = GetSolarStatus(Cyclops);
             float availableSolarEnergy = SolarChargingFactor * solarStatus;
             solarPercentage = solarStatus * 100;
 
-            if (availableSolarEnergy > PowerManager.MinimalPowerValue)
+            if (availableSolarEnergy > MinimalPowerValue)
             {
-                SolarState = SolarState.SunAvailable;
-                float mk1Power = this.SolarChargers.Count * availableSolarEnergy;
-                float mk2Power = this.SolarChargerMk2.Count * availableSolarEnergy * PowerManager.Mk2ChargeRateModifier;
+                solarState = SolarState.SunAvailable;
+                float mk1Power = upgradeHandler[solarTier1] * availableSolarEnergy;
+                float mk2Power = upgradeHandler[solarTier2] * availableSolarEnergy * Mk2ChargeRateModifier;
 
-                this.SolarChargerMk2.RechargeBatteries(mk1Power + mk2Power);
+                upgradeHandler.RechargeBatteries(mk1Power + mk2Power);
 
                 return mk1Power + mk2Power;
```

Comment:
It's time to nerf **Solar** _and_ **Thermal** chargers with diminishing returns as more upgrades are stacked.  
Currently extra modules scale linearly.  
Instead this could be: 
1 => 100%
2 => 100% + 50% = 150%
3 => 100% + 50% + 25% = 175%
4 => 100% + 50% + 25% + 12.5% = 187.5%
5 => 100% + 50% + 25% + 12.5% + 6.25% =193.75%
6 => 100% + 50% + 25% + 12.5% + 6.25% + 3.125% = 196.875%

Reply:
It's a geometric series, so this can be calculated in one step without iterations using this formula:  
(1 - 0.5^N)/(0.5)